# Part 2 : Machine Learning Model Building

In [1]:
import pandas as pd # for data wrangling purpose
import numpy as np # Basic computation library
import seaborn as sns # For Visualization 
import matplotlib.pyplot as plt # ploting package
%matplotlib inline
import warnings # Filtering warnings
warnings.filterwarnings('ignore')

**Importing excel featured engineered Data from Part 1**

In [2]:
df= pd.read_excel('Final ML Data.xlsx')

In [9]:
df.head()

Fuel Type  KMs driven  Engine Displacement(CC)  Transmission  Milage(kmpl)  \
0          4       18600                     1497             1         17.40   
1          1       15000                     1956             1         17.10   
2          1      115000                     2499             1         14.80   
3          4       80000                     1497             1         16.80   
4          4       35000                     1197             1         20.36   

   Max Power(bhp)  Torque(Nm)  Seating Capacity  Color  Gear Box  ...  \
0           117.3       145.0                 5    185         2  ...   
1           170.0       350.0                 5    185         4  ...   
2            80.0        19.0                 7    185         2  ...   
3           116.4       146.0                 5    185         2  ...   
4            78.9       111.8                 5    185         2  ...   

   No of Cylinder  Turbo Charger  Super Charger  Length(mm)  Width(mm)  \
0               4              0              0      4440.0     1695.0   
1               4              2              0      4395.0     1818.0   
2               4              2              0      4435.0     1680.0   
3               4              0              0      4440.0     1695.0   
4               4              0              0      3585.0     1595.0   

   Height(mm)  Price (Rs.)  Car_Age  Car_Brand  Car_Model  
0        1495     675000.0        7         59       1696  
1        1640    1750000.0        4         87        355  
2        1765     175000.0       13         29        712  
3        1485     350000.0       13         59       1326  
4        1550     300000.0        9         78       1437  

[5 rows x 25 columns]

In [10]:
# Splitting data in target and dependent feature
X = df.drop(['Price (Rs.)'], axis = 1)
Y = df['Price (Rs.)']

In [11]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scale = scaler.fit_transform(X)

**Importing require Machine Learning Library**

In [14]:
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import ExtraTreesRegressor
from xgboost import XGBRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.ensemble import  GradientBoostingRegressor
from sklearn.ensemble import BaggingRegressor

In [15]:
from sklearn.metrics import mean_squared_error, mean_absolute_error,  r2_score
from sklearn.model_selection import train_test_split

In [16]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, random_state=42, test_size=.3)
print('Training feature matrix size:',X_train.shape)
print('Training target vector size:',Y_train.shape)
print('Test feature matrix size:',X_test.shape)
print('Test target vector size:',Y_test.shape)

Training feature matrix size: (6218, 24)
Training target vector size: (6218,)
Test feature matrix size: (2666, 24)
Test target vector size: (2666,)


#### **Finding Best Random State**

In [101]:
from sklearn.ensemble import RandomForestRegressor
maxAccu=0
maxRS=0
for i in range(1,200):
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, random_state=i, test_size=.3)
    mod = RandomForestRegressor()
    mod.fit(X_train, Y_train)
    pred = mod.predict(X_test)
    acc=r2_score(Y_test, pred)
    if acc>maxAccu:
        maxAccu=acc
        maxRS=i
print("Best accuracy is ",maxAccu," on Random_state ",maxRS)

Best accuracy is  0.9655888552211401  on Random_state  114


### **Random Forest Regressor Algorithim**

In [104]:
RFR=RandomForestRegressor()
RFR.fit(X_train,Y_train)
pred=RFR.predict(X_test)
R2_score = r2_score(Y_test,pred)*100
print('R2_score:',R2_score)
print('mean_squared_error:',metrics.mean_squared_error(Y_test,pred))
print('mean_absolute_error:',metrics.mean_absolute_error(Y_test,pred))
print('root_mean_squared_error:',np.sqrt(metrics.mean_squared_error(Y_test,pred)))

# Cross Validation Score
scores = cross_val_score(RFR, X, V, cv = 5).mean()*100
print("\nCross validation score :", scores)

# Difference between Accuracy and CV Score
diff = R2_score - scores
print("\nR2_Score - Cross Validation Score :", diff)

R2_score: 96.46493782044703
mean_squared_error: 9226345022.16905
mean_absolute_error: 51153.49883627064
root_mean_squared_error: 96053.86521202076

Cross validation score : 93.03122692981853

R2_Score - Cross Validation Score : 3.433710890628504


### **XGBRegressor ML Model**

In [105]:
XGB=XGBRegressor()
XGB.fit(X_train,Y_train)
pred=XGB.predict(X_test)
R2_score = r2_score(Y_test,pred)*100
print('R2_score:',R2_score)
print('mean_squared_error:',metrics.mean_squared_error(Y_test,pred))
print('mean_absolute_error:',metrics.mean_absolute_error(Y_test,pred))
print('root_mean_squared_error:',np.sqrt(metrics.mean_squared_error(Y_test,pred)))

# Cross Validation on XGB Model
scores = cross_val_score(XGB, X, Y, cv = 5).mean()*100
print("\nCross validation score :", scores)

# Difference between Accuracy and CV Score
diff = R2_score - scores
print("\nR2_Score - Cross Validation Score :", diff)

R2_score: 96.8578288022264
mean_squared_error: 8200918149.917081
mean_absolute_error: 50118.93210268505
root_mean_squared_error: 90558.92087429643

Cross validation score : 93.2469040953667

R2_Score - Cross Validation Score : 3.6109247068596915


### **Gradient Boosting Regressor  ML Model**

In [107]:
GBR=GradientBoostingRegressor()
GBR.fit(X_train,y_train)
pred=GBR.predict(X_test)
R2_score = r2_score(y_test,pred)*100
print('R2_score:',R2_score)
print('mean_squared_error:',metrics.mean_squared_error(Y_test,pred))
print('mean_absolute_error:',metrics.mean_absolute_error(Y_test,pred))
print('root_mean_squared_error:',np.sqrt(metrics.mean_squared_error(Y_test,pred)))

# Cross Validation on Gradient Boosting
scores = cross_val_score(GBR, X, Y, cv = 5).mean()*100
print("\nCross validation score :", scores)

# Difference between Accuracy and CV Score
diff = R2_score - scores
print("\nR2_Score - Cross Validation Score :", diff)

R2_score: 94.9328623763045
mean_squared_error: 13224989439.06563
mean_absolute_error: 71240.04884627696
root_mean_squared_error: 114999.95408288487

Cross validation score : 90.1937305617025

R2_Score - Cross Validation Score : 4.739131814602004


###  **Decision Tree Regressor ML Model**

In [108]:
DTR=DecisionTreeRegressor()
DTR.fit(X_train,y_train)
pred=DTR.predict(X_test)
R2_score = r2_score(y_test,pred)*100
print('R2_score:',R2_score)
print('mean_squared_error:',metrics.mean_squared_error(Y_test,pred))
print('mean_absolute_error:',metrics.mean_absolute_error(Y_test,pred))
print('root_mean_squared_error:',np.sqrt(metrics.mean_squared_error(Y_test,pred)))
# Cross Validation Score
scores = cross_val_score(DTR, X, Y, cv = 5).mean()*100
print("\nCross validation score :", scores)

# Difference between Accuracy and CV Score
diff = R2_score - scores
print("\nR2_Score - Cross Validation Score :", diff)

R2_score: 91.79408304824462
mean_squared_error: 21417054969.521046
mean_absolute_error: 64770.82728592162
root_mean_squared_error: 146345.6694594037

Cross validation score : 88.83907795864332

R2_Score - Cross Validation Score : 2.9550050896013005


### **Bagging Regressor ML Model**

In [109]:
BR=BaggingRegressor()
BR.fit(X_train,Y_train)
pred=BR.predict(X_test)
R2_score = r2_score(y_test,pred)*100
print('R2_score:',R2_score)
print('mean_squared_error:',metrics.mean_squared_error(Y_test,pred))
print('mean_absolute_error:',metrics.mean_absolute_error(Y_test,pred))
print('root_mean_squared_error:',np.sqrt(metrics.mean_squared_error(Y_test,pred)))

# Cross Validation Score
scores = cross_val_score(BR, X, Y, cv = 5).mean()*100
print("\nCross validation score :", scores)

# Difference between Accuracy and CV Score
diff = R2_score - scores
print("\nR2_Score - Cross Validation Score :", diff)

R2_score: 95.67561886178403
mean_squared_error: 11286430156.537165
mean_absolute_error: 55118.9760619255
root_mean_squared_error: 106237.61177914894

Cross validation score : 92.60347410600774

R2_Score - Cross Validation Score : 3.072144755776293


### Final model Selection
 **All model are giving us R2 score & Cross validation Score more than 90%, So we will select model which has less difference between these score.**
 
**<u>On Basis of difference between R2 Score and Cross Validation Score Decision Tree Regressor</u> is selected as best model with 91.79% r2_score.**

**We will perform Hyper Parameter tuning over this model**

## **Hyper Parameter Tunning** 

In [110]:
#importing necessary libraries
from sklearn.model_selection import GridSearchCV

In [111]:
parameter = {'criterion':['squared_error', 'friedman_mse', 'absolute_error', 'poisson'],
             'splitter':['best','random'],
             'max_features':['auto','sqrt','log2'],
             'min_samples_split':[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15],
             'max_depth':[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]}

Giving DecisionTreeRegressor parameters.

In [112]:
GCV=GridSearchCV(DecisionTreeRegressor(),parameter,cv=5)

Running grid search CV for decisionTreesRegressor.

In [113]:
DecisionTreeRegressorGCV.fit(X_train,y_train)

GridSearchCV(cv=10, estimator=DecisionTreeRegressor(),
             param_grid={'criterion': ['squared_error', 'friedman_mse',
                                       'absolute_error', 'poisson'],
                         'max_depth': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12,
                                       13, 14, 15],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'min_samples_split': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
                                               11, 12, 13, 14, 15],
                         'splitter': ['best', 'random']})

Tunning the model using GCV.

In [114]:
GCV.best_params_

{'criterion': 'friedman_mse',
 'max_depth': 13,
 'max_features': 'auto',
 'min_samples_split': 4,
 'splitter': 'random'}

## **Final Model**

In [115]:
Final_mod=DecisionTreeRegressor(criterion='friedman_mse', max_depth=15, max_features='auto',
                                min_samples_split=4, splitter='random')
Final_mod.fit(X_train,Y_train)
pred=Final_mod.predict(X_test)
print('R2_Score:',r2_score(Y_test,pred)*100)
print('mean_squared_error:',metrics.mean_squared_error(Y_test,pred))
print('mean_absolute_error:',metrics.mean_absolute_error(Y_test,pred))
print("RMSE value:",np.sqrt(metrics.mean_squared_error(Y_test, pred)))

R2_Score: 92.28906442588051
mean_squared_error: 20125176994.634956
mean_absolute_error: 70467.88619495885
RMSE value: 141863.2334138587


#### **Final Model is giving us R2 Score of 92.29% which is slightly improved compare to earlier R2 score of 91.79%.**

## **Saving the model**

In [116]:
# Saving the model using .pkl
import joblib
joblib.dump(Final_mod,"Car_Price.pkl")

['Car_Price.pkl']

## **Predictions Using Final Model**

In [117]:
# Loading the saved model
model=joblib.load("Car_Price.pkl")

#Prediction
prediction = model.predict(X_test)
prediction

array([ 379000.        , 1650000.        ,  355000.        , ...,
        556333.33333333,  199923.07692308,  364800.        ])

In [118]:
pd.DataFrame([model.predict(X_test)[:],y_test[:]],index=["Predicted","Actual"])

0          1         2              3              4     \
Predicted  379000.0  1650000.0  355000.0  332666.666667  490333.333333   
Actual     379000.0  1650000.0  465000.0  435000.000000  550000.000000   

                    5         6         7             8         9     \
Predicted  470382.352941  590000.0  590000.0  1.541333e+06  687000.0   
Actual     450000.000000  550000.0  550000.0  1.500000e+06  643000.0   

                   10             11        12             13    \
Predicted  7.133333e+05  412017.241379  327500.0  684333.333333   
Actual     1.125000e+06  385000.000000  325000.0  705000.000000   

                    14             15        16         17        18    \
Predicted  273333.333333  513666.666667  262500.0  1850000.0  550800.0   
Actual     275000.000000  485000.000000  265000.0  1630000.0  550000.0   

               19             20        21        22             23    \
Predicted  667500.0  726666.666667  925000.0  630500.0  373428.571429   
Actual     549000.0  640000.000000  365000.0  553000.0  424000.000000   

               24             25        26             27        28    \
Predicted  447500.0  518666.666667  500000.0  414785.714286   97550.0   
Actual     410000.0  543000.000000  399000.0  325000.000000  161000.0   

               29        30             31        32             33    \
Predicted  609000.0  807000.0  382558.139535  637000.0  364735.849057   
Actual     613000.0  940000.0  299000.000000  550000.0  345000.000000   

               34        35        36        37        38             39    \
Predicted  515000.0  282000.0  323000.0  906000.0  255000.0  364666.666667   
Actual     640000.0  250000.0  429000.0  915000.0  230000.0  392000.000000   

                    40        41             42            43         44    \
Predicted  799666.666667  255000.0  540666.666667  1.357833e+06  2500000.0   
Actual     864000.000000  335000.0  536000.000000  1.251000e+06  3200000.0   

               45             46             47             48    \
Predicted  437500.0  316948.275862  422666.666667  520411.764706   
Actual     440000.0  255000.000000  460000.000000  500000.000000   

                    49        50        51             52             53    \
Predicted  680666.666667  536000.0  388000.0  274333.333333  729666.666667   
Actual     875000.000000  536000.0  317000.0  283000.000000  565000.000000   

               54             55             56             57    \
Predicted  452500.0  516952.380952  379444.444444  217666.666667   
Actual     399000.0  472000.000000  426000.000000  295000.000000   

                    58            59             60        61        62    \
Predicted  458738.095238  343335.57047  722703.703704  312500.0  922500.0   
Actual     370000.000000  363000.00000  825000.000000  240000.0  950000.0   

               63             64        65        66             67    \
Predicted  536500.0  675465.116279  618000.0  650000.0  497631.578947   
Actual     805000.0  576000.000000  590000.0  708000.0  460000.000000   

                    68         69             70             71        72    \
Predicted  528333.333333  1222500.0  321666.666667  568294.117647  486000.0   
Actual     446000.000000  1375000.0  410000.000000  553000.000000  486000.0   

               73        74         75        76        77             78    \
Predicted  490000.0  609000.0  1150000.0  657000.0  857500.0  210047.619048   
Actual     480000.0  734000.0   875000.0  565000.0  900000.0  190000.000000   

                    79        80        81        82         83    \
Predicted  643333.333333  747750.0  287000.0  696700.0  1087500.0   
Actual     375000.000000  670000.0  250000.0  776000.0  1075000.0   

                    84        85        86             87        88    \
Predicted  364735.849057  319750.0  530000.0  568333.333333  405000.0   
Actual     300000.000000  234000.0  450000.0  735000.000000  249000.0   

                89 